File: backend/app/schemas/fixed_deposit.py
from datetime import date, datetime

from pydantic import BaseModel


class FixedDepositBase(BaseModel):
    bank_name: str
    principal: float
    interest_rate: float
    compounding_frequency: str = "quarterly"
    start_date: date
    maturity_date: date
    maturity_amount: float | None = None
    is_cumulative: bool = True
    interest_payout_freq: str | None = None
    auto_renew: bool = False
    notes: str | None = None


class FixedDepositCreate(FixedDepositBase):
    pass


class FixedDepositUpdate(BaseModel):
    bank_name: str | None = None
    principal: float | None = None
    interest_rate: float | None = None
    compounding_frequency: str | None = None
    start_date: date | None = None
    maturity_date: date | None = None
    maturity_amount: float | None = None
    is_cumulative: bool | None = None
    interest_payout_freq: str | None = None
    auto_renew: bool | None = None
    notes: str | None = None


class FixedDepositResponse(FixedDepositBase):
    id: int
    current_value: float | None = None
    status: str = "active"
    closure_date: date | None = None
    closure_amount: float | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class FixedDepositWithTags(FixedDepositResponse):
    tags: list["TagBrief"] = []


class TagBrief(BaseModel):
    id: int
    name: str
    color: str | None = None

    model_config = {"from_attributes": True}


class FDTagAssign(BaseModel):
    tag_ids: list[int]


class FDCloseRequest(BaseModel):
    closure_date: date
    closure_amount: float
    premature: bool = False
    notes: str | None = None


class FDRenewRequest(BaseModel):
    new_maturity_date: date
    new_principal: float | None = None
    new_interest_rate: float | None = None
    new_compounding_frequency: str | None = None
    new_bank_name: str | None = None
    new_is_cumulative: bool | None = None
    notes: str | None = None


class FDRenewResponse(BaseModel):
    closed_fd: FixedDepositResponse
    new_fd: FixedDepositResponse


FixedDepositWithTags.model_rebuild()

File: frontend/src/types/fixed-deposit.ts
/** Matches backend FixedDepositResponse (no tags) */
export interface FixedDepositResponse {
  id: number;
  bank_name: string;
  principal: number;
  interest_rate: number;
  compounding_frequency: string;
  start_date: string;
  maturity_date: string;
  maturity_amount: number | null;
  current_value: number | null;
  is_cumulative: boolean;
  interest_payout_freq: string | null;
  auto_renew: boolean;
  funded_externally: boolean;
  notes: string | null;
  status: "active" | "closed";
  closure_date: string | null;
  closure_amount: number | null;
  created_at: string;
  updated_at: string;
}

/** Matches backend FixedDepositWithTags (list/detail endpoint, includes tags) */
export interface FixedDeposit extends FixedDepositResponse {
  tags: FDTagBrief[];
}

export interface FDTagBrief {
  id: number;
  name: string;
  color: string | null;
}

export interface FixedDepositCreate {
  bank_name: string;
  principal: number;
  interest_rate: number;
  compounding_frequency: string;
  start_date: string;
  maturity_date: string;
  maturity_amount?: number;
  is_cumulative?: boolean;
  interest_payout_freq?: string;
  auto_renew?: boolean;
  funded_externally?: boolean;
  notes?: string;
}

export interface FixedDepositUpdate {
  bank_name?: string;
  principal?: number;
  interest_rate?: number;
  compounding_frequency?: string;
  start_date?: string;
  maturity_date?: string;
  maturity_amount?: number;
  is_cumulative?: boolean;
  interest_payout_freq?: string;
  auto_renew?: boolean;
  notes?: string;
}

export type FDStatus = "active" | "maturing_soon" | "matured" | "closed";

export function getFDStatus(fd: FixedDeposit): FDStatus {
  if (fd.status === "closed") return "closed";

  const now = new Date();
  const maturity = new Date(fd.maturity_date);
  const daysToMaturity = Math.ceil(
    (maturity.getTime() - now.getTime()) / (1000 * 60 * 60 * 24)
  );

  if (daysToMaturity <= 0) return "matured";
  if (daysToMaturity <= 30) return "maturing_soon";
  return "active";
}

export interface FDCloseRequest {
  closure_date: string;
  closure_amount: number;
  premature: boolean;
  reinvesting?: boolean;
  notes?: string;
}

export interface FDRenewRequest {
  new_maturity_date: string;
  new_principal?: number;
  new_interest_rate?: number;
  new_compounding_frequency?: string;
  new_bank_name?: string;
  new_is_cumulative?: boolean;
  notes?: string;
}

export interface FDRenewResponse {
  closed_fd: FixedDepositResponse;
  new_fd: FixedDepositResponse;
}
